pkg/provisioner/ironic/delete_test.go

```diff
@@ -190,3 +219,100 @@ func TestDelete(t *testing.T) {
 		})
 	}
 }
+
+func TestDeleteWithPowerOff(t *testing.T) {
+
+	nodeUUID := "33ce8659-7400-4c68-9535-d10766f07a58"
+
+	cases := []struct {
+		name        string
+		ironic      *testserver.IronicMock
+		inspector   *testserver.InspectorMock
+		priorErrors int
+
+		expectedPowerState string
+		expectedError      string
+	}{
+		{
+			name: "test-power-change-without-errors",
+			ironic: testserver.NewIronic(t).Node(
+				nodes.Node{
+					UUID:           nodeUUID,
+					ProvisionState: "active",
+					Maintenance:    true,
+					PowerState:     powerOn,
+				},
+			).WithNodeStatesPowerUpdate(nodeUUID, http.StatusAccepted),
+			priorErrors:        0,
+			expectedPowerState: "power off",
+			expectedError:      "failed to remove host",
+		},
+		{
+			name: "test-power-change-with-errors",
+			ironic: testserver.NewIronic(t).Node(
+				nodes.Node{
+					UUID:           nodeUUID,
+					ProvisionState: "active",
+					Maintenance:    true,
+					PowerState:     powerOn,
+				},
+			),
+			priorErrors:        2,
+			expectedPowerState: "power off",
+			expectedError:      "",
+		},
+		{
+			name: "test-power-change-with-max-errors",
+			ironic: testserver.NewIronic(t).Node(
+				nodes.Node{
+					UUID:           nodeUUID,
+					ProvisionState: "active",
+					Maintenance:    true,
+					PowerState:     powerOn,
+				},
+			),
+			priorErrors:        4,
+			expectedPowerState: "",
+			expectedError:      "failed to remove host",
+		},
+	}
```

Comment:
A couple of another cases are required to manage when `p.hardPowerOff()` returns an error

Reply:
I wasn't able to control the return value of hardPowerOff() using  WithNodeStatesPowerUpdate() in https://github.com/metal3-io/baremetal-operator/blob/master/pkg/provisioner/ironic/testserver/ironic.go#L227 .

Any ideas. Only the mock for Delete() seems to be affecting the outcome.

Reply:
Using https://github.com/metal3-io/baremetal-operator/blob/master/pkg/provisioner/ironic/testserver/ironic.go#L227, I wasn't able to mock the out come for hardPowerOff() which is called from within the Delete().

Reply:
That's a real interesting case and it's worth some additional explanation.

Adding the mocking function to test case should be ok, ie: 
```
ironic: testserver.NewIronic(t).Node(
				nodes.Node{
					UUID:           nodeUUID,
					ProvisionState: "active",
					Maintenance:    true,
					PowerState:     powerOn,
				},
			).WithNodeStatesPowerUpdate(nodeUUID, http.StatusConflict).Delete(nodeUUID),
```
In this case the expected error is returned, anyhow this type assertion https://github.com/sadasu/baremetal-operator/blob/4d0f31405af55c03516533cd6e624aa6c32afaa2/pkg/provisioner/ironic/ironic.go#L1624  is going to fail because the real error is wrapped (twice).

As per the golang recommended best practices on errors (https://blog.golang.org/go1.13-errors), type check must be changed using the new errors functions like: 

```
var hostErr *HostLockedError
if errors.As(err, &hostErr) {
  p.log.Info("could not power off host, busy")
  return retryAfterDelay(powerRequeueDelay)
} else {
  return operationFailed("failed to power off host")
}
```

But that's not _yet_ sufficient, because the underlying `changePower` does not return a pointer receiver for the error: https://github.com/sadasu/baremetal-operator/blob/4d0f31405af55c03516533cd6e624aa6c32afaa2/pkg/provisioner/ironic/ironic.go#L1681

So that code as well must be changed to:

```
return result, &HostLockedError{Address: p.host.Spec.BMC.Address}
```
This means of course that other points in the code (ie `PowerOn`) needs to be reviewed for properly managing the error

Reply:
I have attempt to do this in https://github.com/metal3-io/baremetal-operator/pull/816/commits/c941bc0ddd6324299b68272855578b1b38c879ba. This commit might need more work.